File: extensions/wordflow/engine/circuit_breaker.py
# -*- coding: utf-8 -*-
"""CircuitBreaker — T19. CLOSED / OPEN / HALF_OPEN. 0% LLM."""
from __future__ import annotations

import time
from typing import Any, Callable


class CircuitOpenError(Exception):
    def __init__(self, name: str, retry_at: float):
        self.name = name
        self.retry_at = retry_at
        super().__init__(f"circuit open: {name}")


class CircuitBreaker:
    """Classic breaker. clock injectable for tests."""

    def __init__(
        self,
        name: str = "default",
        *,
        failure_threshold: int = 3,
        recovery_timeout_s: float = 30.0,
        success_threshold: int = 1,
        clock: Any = None,
    ):
        if failure_threshold < 1:
            raise ValueError("failure_threshold >= 1")
        if recovery_timeout_s <= 0:
            raise ValueError("recovery_timeout_s > 0")
        self.name = name
        self.failure_threshold = failure_threshold
        self.recovery_timeout_s = float(recovery_timeout_s)
        self.success_threshold = max(1, success_threshold)
        self._clock = clock or time.monotonic
        self.state = "CLOSED"  # CLOSED | OPEN | HALF_OPEN
        self.failure_count = 0
        self.success_count = 0
        self.opened_at: float | None = None

    def _now(self) -> float:
        return float(self._clock())

    def allow(self) -> bool:
        if self.state == "CLOSED":
            return True
        if self.state == "OPEN":
            assert self.opened_at is not None
            if self._now() - self.opened_at >= self.recovery_timeout_s:
                self.state = "HALF_OPEN"
                self.success_count = 0
                return True
            return False
        # HALF_OPEN: allow probe
        return True

    def record_success(self) -> None:
        if self.state == "HALF_OPEN":
            self.success_count += 1
            if self.success_count >= self.success_threshold:
                self.state = "CLOSED"
                self.failure_count = 0
                self.opened_at = None
        elif self.state == "CLOSED":
            self.failure_count = 0

    def record_failure(self) -> None:
        if self.state == "HALF_OPEN":
            self.state = "OPEN"
            self.opened_at = self._now()
            self.failure_count = self.failure_threshold
            self.success_count = 0
            return
        self.failure_count += 1
        if self.failure_count >= self.failure_threshold:
            self.state = "OPEN"
            self.opened_at = self._now()

    def call(self, fn: Callable[[], Any]) -> Any:
        if not self.allow():
            retry_at = (self.opened_at or 0) + self.recovery_timeout_s
            raise CircuitOpenError(self.name, retry_at)
        try:
            result = fn()
        except Exception:
            self.record_failure()
            raise
        self.record_success()
        return result

    def snapshot(self) -> dict[str, Any]:
        return {
            "name": self.name,
            "state": self.state,
            "failure_count": self.failure_count,
            "success_count": self.success_count,
            "opened_at": self.opened_at,
            "failure_threshold": self.failure_threshold,
            "recovery_timeout_s": self.recovery_timeout_s,
        }

File: Core kernel Yaiwes/Oxios/code/crates/oxios-kernel/src/a2a/circuit_breaker.rs
//! A2A protocol circuit breaker for delegation reliability.
//!
//! Prevents cascading failures when A2A delegation repeatedly fails.
//!
//! # Example
//!
//! ```
//! use oxios_kernel::a2a::circuit_breaker::{A2ACircuitBreaker, CircuitState};
//!
//! let cb = A2ACircuitBreaker::new(3, 30);  // 3 failures, 30s reset
//! assert_eq!(cb.state(), CircuitState::Closed);
//! assert!(cb.is_allowed());
//! ```

use std::sync::atomic::{AtomicU8, AtomicU32, AtomicU64, Ordering};
use std::time::{Duration, SystemTime, UNIX_EPOCH};

/// Current wall-clock time in seconds since the UNIX epoch. Stored in the
/// breaker's `AtomicU64` because `Instant` cannot be encoded as u64. The
/// previous implementation stored `Instant::now().elapsed().as_secs()`, which
/// is always 0 — so once Open the breaker never recovered.
fn now_epoch_secs() -> u64 {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map(|d| d.as_secs())
        .unwrap_or(0)
}

/// Circuit breaker states.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum CircuitState {
    /// Normal operation — requests allowed.
    Closed,
    /// Too many failures — requests blocked.
    Open,
    /// Testing recovery — limited requests allowed.
    HalfOpen,
}

impl CircuitState {
    fn from_u8(v: u8) -> Self {
        match v {
            0 => CircuitState::Closed,
            1 => CircuitState::Open,
            2 => CircuitState::HalfOpen,
            _ => CircuitState::Closed,
        }
    }
}

/// A2A delegation circuit breaker.
///
/// Tracks failures and opens the circuit when threshold is exceeded.
/// After timeout, allows limited test requests (half-open state).
#[derive(Debug)]
pub struct A2ACircuitBreaker {
    state: AtomicU8,
    failure_count: AtomicU32,
    success_count: AtomicU32,
    last_failure_time: AtomicU64,
    threshold: u32,
    reset_timeout: Duration,
}

impl A2ACircuitBreaker {
    /// Create a new circuit breaker.
    ///
    /// # Arguments
    /// * `threshold` - Number of consecutive failures before opening
    /// * `reset_timeout_secs` - Seconds to wait before testing recovery
    pub fn new(threshold: u32, reset_timeout_secs: u64) -> Self {
        Self {
            state: AtomicU8::new(CircuitState::Closed as u8),
            failure_count: AtomicU32::new(0),
            success_count: AtomicU32::new(0),
            last_failure_time: AtomicU64::new(0),
            threshold,
            reset_timeout: Duration::from_secs(reset_timeout_secs),
        }
    }

    /// Current circuit state.
    ///
    /// `SeqCst` so a freshly-OPEN breaker is visible to every thread that
    /// gates request flow on this read (audit F-15 — `Relaxed` could let a
    /// burst of requests past a transition that hasn't propagated yet).
    pub fn state(&self) -> CircuitState {
        CircuitState::from_u8(self.state.load(Ordering::SeqCst))
    }

    /// Whether a request is allowed through the circuit.
    pub fn is_allowed(&self) -> bool {
        match self.state() {
            CircuitState::Closed => true,
            CircuitState::Open => {
                // Check if reset timeout has passed
                let last_failure = self.last_failure_time.load(Ordering::Acquire);
                let elapsed = now_epoch_secs().saturating_sub(last_failure);
                if elapsed > self.reset_timeout.as_secs() {
                    // Transition to half-open.
                    self.state
                        .store(CircuitState::HalfOpen as u8, Ordering::SeqCst);
                    self.success_count.store(0, Ordering::Release);
                    true
                } else {
                    false
                }
            }
            CircuitState::HalfOpen => {
                // Allow limited requests (up to 2)
                self.success_count.load(Ordering::Acquire) < 2
            }
        }
    }

    /// Record a successful request.
    pub fn record_success(&self) {
        match self.state() {
            CircuitState::HalfOpen => {
                let successes = self.success_count.fetch_add(1, Ordering::AcqRel) + 1;
                if successes >= 2 {
                    // Recovery successful → closed
                    self.state
                        .store(CircuitState::Closed as u8, Ordering::SeqCst);
                    self.failure_count.store(0, Ordering::Release);
                    tracing::info!("A2A circuit breaker CLOSED (recovery successful)");
                }
            }
            CircuitState::Closed => {
                // Reset failure count on success
                self.failure_count.store(0, Ordering::Release);
            }
            CircuitState::Open => {}
        }
    }

    /// Record a failed request.
    pub fn record_failure(&self) {
        let failures = self.failure_count.fetch_add(1, Ordering::AcqRel) + 1;
        self.last_failure_time
            .store(now_epoch_secs(), Ordering::Release);

        if failures >= self.threshold && self.state() != CircuitState::Open {
            self.state.store(CircuitState::Open as u8, Ordering::SeqCst);
            tracing::warn!(
                failures,
                threshold = self.threshold,
                "A2A circuit breaker OPEN"
            );
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_initial_state_is_closed() {
        let cb = A2ACircuitBreaker::new(3, 10);
        assert_eq!(cb.state(), CircuitState::Closed);
        assert!(cb.is_allowed());
    }

    #[test]
    fn test_opens_after_threshold() {
        let cb = A2ACircuitBreaker::new(3, 10);

        cb.record_failure();
        assert_eq!(cb.state(), CircuitState::Closed);

        cb.record_failure();
        assert_eq!(cb.state(), CircuitState::Closed);

        cb.record_failure(); // Now at threshold
        assert_eq!(cb.state(), CircuitState::Open);
        assert!(!cb.is_allowed());
    }

    #[test]
    fn test_success_resets_failure_count() {
        let cb = A2ACircuitBreaker::new(3, 10);

        cb.record_failure();
        cb.record_failure();
        cb.record_success(); // Should reset

        assert_eq!(cb.failure_count.load(Ordering::Relaxed), 0);
    }
}
